#' Get filepaths from GDCquery
#'
#' Gets the filepath to which the raw data will be downloaded from the GDCquery created
#'by TCGAbiolinks.
#'
#' @param query GDCquery object generated by TCGAbiolinks::GDCquery
#' @return Character vector of the pathway to the GDCfolder
#' @example
#' getQueryFilePaths(query)
#' getQueryFilePaths(GDCquery(project, data.category = "Transcriptome Profiling",
#' data.type = "Gene Expression Quantification", experimental.strategy = "RNA-Seq",
#' workflow.type = "HTSeq - Counts", barcodes)

getQueryFilePaths <- function(query){

  source <- ifelse(query$legacy, "legacy", "harmonized")

  filepath <- paste(query$results[[1]]$project[1],
                    source,
                    gsub(" ", "_", query$results[[1]]$data_category[1]),
                    gsub(" ", "_", query$results[[1]]$data_type[1]), sep = "/")

  filepath
}

#' Download gene expression data from the TCGA
#'
#' Downloads expression data (RNA-seq or microarray) for the specified project in the TCGA.
#'
#' @param project Name of a TCGA project (eg. TCGA-PRAD or TCGA-BRCA).
#' @param barcodes Vector containing the sample barcodes to download.
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @param microarray Wether to check for available microarray data (in the TCGA legacy) or RNA-seq
#' counts (in the harmonised TCGA). Defaults to FALSE.
#' @return Downloads a CSV file per sample containing the expression data and creates a merged CSV with all
#' a sample in each column.
#' @examples
#' downloadExpression(project = "TCGA-CHOL", barcodes = c("TCGA-W5-AA2X-01A", "TCGA-ZU-A8S4-11A"),
#' microarray = FALSE)
#' @note Barcodes need to contain the sample code (01A or 11A among others). An easy way to get barcodes
#' is using the function getBarcodes.
#' @importFrom TCGAbiolinks GDCquery getResults GDCprepare GDCdownload
#' @importFrom SummarizedExperiment assay
#' @export

downloadExpression <- function(project, barcodes, automataID, microarray = FALSE){

  # Checks the microarray argument
  if(microarray == FALSE){
    type = "RNA-seq"
  } else if (microarray == TRUE){
    type = "microarray"
  } else{
    stop("Please, set TRUE or FALSE for microarray.")
  }

  # Check if the file already exists

  if(file.exists(here::here("TCGA_results", project, automataID, "TCGA_data",
                            sprintf("Expression_matrix_%s_%s_%s.csv", type, project, automataID))) == TRUE){

    message("---------------------------------")
    message("~ DOWNLOADING EXPRESSION DATA ~")
    message("---------------------------------")
    message("o Data has been already downloaded and processed.")
    message("o Skipping...")

  } else{

    # Creates folder to save the data
    project <- toupper(project)
    createTCGAfolder(project, automataID)


    # Downloads the data

    if(microarray == FALSE){

      type = "RNA-seq"

      message("----------------------------------------")
      message("~ DOWNLOADING RNA-SEQ EXPRESSION DATA ~")
      message("----------------------------------------")

      query = GDCquery(project = project,
                       data.category = "Transcriptome Profiling",
                       data.type = "Gene Expression Quantification",
                       experimental.strategy = "RNA-Seq",
                       workflow.type = "HTSeq - Counts",
                       barcode = barcodes)

    } else if (microarray == TRUE){

      type = "microarray"

      message("---------------------------------------------")
      message("~ DOWNLOADING MICROARRAY EXPRESSION DATA ~")
      message("---------------------------------------------")

      query = GDCquery(project = project,
                       data.category = "Gene expression",
                       data.type = "Gene expression quantification",
                       experimental.strategy = "Gene expression array",
                       barcode = barcodes,
                       legacy = TRUE)

    } else{
      stop("Please, set TRUE or FALSE for microarray.")
    }

    # Generates a table data of the found files
    expression_files = getResults(query)

    # Downloads the files (it will create a series of subfolders in the working directory automatically)
    message("")
    message("")
    message("--------------------")
    message("o Downloading data")
    message("--------------------")
    GDCdownload(query,
                method = "api",
                files.per.chunk = 70)

    filepath <- getQueryFilePaths(query)
    message(paste("Raw data is saved in ", filepath))

    message("--------------------")
    message("oo Preparing data")
    message("--------------------")
    SE_expression = GDCprepare(query = query, summarizedExperiment = FALSE)


    # Extracts the expression matrix
    message("-------------------------------------")
    message("ooo Generating expression matrix...")
    message("-------------------------------------")
    expression_matrix <- SE_expression %>%
      as.data.frame() %>%
      mutate(rowNames = row.names(.)) %>%
      distinct()

    # So TCGAbiolinks or the GDC has changed how the colnames are called which is causing me tons of problems

    if(type == "microarray"){

      rownames(expression_matrix) <- expression_matrix$`Composite Element REF`

      expression_matrix <- select(expression_matrix, -c(`Composite Element REF`, rowNames))
      colnames(expression_matrix) <- gsub("log2_lowess_normalized__cy5_cy3__collapsed_by_gene_symbol_", "", colnames(expression_matrix))

      write.csv(expression_matrix, here::here("TCGA_results", project, automataID, "TCGA_data",
                                              sprintf("Expression_matrix_%s_%s_%s.csv", type, project, automataID)), row.names = TRUE)
      message(paste("Expression matrix has been saved in",
                    here::here("TCGA_results", project, automataID, "TCGA_data",
                               sprintf("Expression_matrix_%s_%s_%s.csv", type, project, automataID))))

    } else {

      rownames(expression_matrix) <- expression_matrix$rowNames

      expression_matrix <- select(expression_matrix, -c(rowNames)) %>%
        filter(!grepl("__", X1))

      write.csv(expression_matrix, here::here("TCGA_results", project, automataID, "TCGA_data",
                                              sprintf("Expression_matrix_%s_%s_%s.csv", type, project, automataID)), row.names = FALSE)
      message(paste("Expression matrix has been saved in",
                    here::here("TCGA_results", project, automataID, "TCGA_data",
                               sprintf("Expression_matrix_%s_%s_%s.csv", type, project, automataID))))
    }



    write.csv(expression_files, here::here("TCGA_results", project, automataID, "file_information",
                                           sprintf("Expression_files_%s_%s_%s.csv", type, project, automataID)))
    message(paste("File information has been saved in",
                  here::here("TCGA_results", project, automataID, "file_information",
                             sprintf("Expression_files_%s_%s_%s.csv", type, project, automataID))))

  }

}

#' Download clinical data from the TCGA
#'
#' Downloads clinical data for the specified project in the TCGA.
#'
#' @param project Name of a TCGA project (eg. TCGA-PRAD or TCGA-BRCA).
#' @param barcodes Vector containing participant barcodes for the cases to download.
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @return Downloads a CSV file per sample containing the clinical data and creates a merged CSV with all
#' a sample in each column.
#' @note Barcodes need to contain the participant code (eg. TCGA-WS-AA2X). An easy way to get barcodes
#' is using the function getBarcodes.
#' @importFrom TCGAbiolinks GDCprepare_clinic
#' @export

downloadClinical <- function(project, barcodes, automataID){


  message("--------------------------------")
  message("~ DOWNLOADING CLINICAL DATA ~")
  message("--------------------------------")

  if(file.exists(here::here("TCGA_results", project, automataID, "TCGA_data",
                            sprintf("clinical_data_%s_%s.csv", project, automataID))) == FALSE){

    # Creates folder to save the data
    project <- toupper(project)
    createTCGAfolder(project, automataID)

    # Downloads the data
    query <- GDCquery(project = project,
                      data.category = "Clinical",
                      file.type = "xml",
                      barcode = barcodes)

    # Generates a table data of the found files
    clinical_files <- getResults(query)

    # Downloads the files (it will create a series of subfolders in the working directory automatically)
    message("")
    message("")
    message("--------------------")
    message("o Downloading data")
    message("--------------------")
    GDCdownload(query,
                method = "api",
                files.per.chunk = 70)

    filepath <- getQueryFilePaths(query)
    message(paste("Raw data is saved in ", filepath))

    message("--------------------")
    message("oo Preparing data")
    message("--------------------")
    prep_clinical <- GDCprepare_clinic(query, clinical.info = "patient")
    prep_clinical <- unique(prep_clinical)

    # Generates output
    message("---------------------------")
    message("ooo Generating output...")
    message("---------------------------")


    write.csv(clinical_files, here::here("TCGA_results", project, automataID, "file_information",
                                         sprintf("clinical_files_%s_%s.csv", project, automataID)))
    message(paste("File information has been saved in",
                  here::here("TCGA_results", project, automataID, "file_information",
                             sprintf("clinical_files_%s.csv", project, automataID))))

    write.csv(prep_clinical, here::here("TCGA_results", project, automataID, "TCGA_data",
                                        sprintf("clinical_data_%s_%s.csv", project, automataID)))
    message(paste("Clinical data has been saved in",
                  here::here("TCGA_results", project, automataID, "TCGA_data",
                             sprintf("clinical_data_%s_%s.csv", project, automataID))))

  } else{
    message("o Data has been already downloaded and processed.")
    message("o Skipping...")
  }

}

#' Download copy number variation (CNV) data from the TCGA
#'
#' Downloads CNV data for the specified project in the TCGA.
#'
#' @param project Name of a TCGA project (eg. TCGA-PRAD or TCGA-BRCA).
#' @param barcodes Vector containing sample barcodes for the cases to download.
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @return Downloads a CSV file per sample containing the CNV data and creates a merged CSV with all
#' a sample in each column.
#' @importFrom TCGAbiolinks GDCquery getResults GDCprepare GDCdownload
#' @examples
#' downloadCNV(project = "TCGA-CHOL", barcodes = c("TCGA-W5-AA2X-01A", "TCGA-ZU-A8S4-11A"))
#' @note Barcodes need to contain the sample code (01A or 11A among others). An easy way to get barcodes
#' is using the function getBarcodes.
#' @export

downloadCNV <- function(project, barcodes, automataID){

  message("--------------------------")
  message("~ DOWNLOADING CNV DATA ~")
  message("--------------------------")

  if(file.exists(here::here("TCGA_results", project, automataID, "TCGA_data",
                            sprintf("CNV_data_%s_%s.csv", project, automataID))) == FALSE){

    # Creates folder to save the data
    project <- toupper(project)
    createTCGAfolder(project, automataID)

    # Downloads the data
    query <- GDCquery(project = project,
                      data.category = "Copy Number Variation",
                      data.type = "Copy Number Segment",
                      barcode = barcodes)

    # Generates a table data of the found files
    CNV_files <- getResults(query)

    # Downloads the files (it will create a series of subfolders in the working directory automatically)
    message("")
    message("")
    message("--------------------")
    message("o Downloading data")
    message("--------------------")
    GDCdownload(query,
                method = "api",
                files.per.chunk = 25)

    filepath <- getQueryFilePaths(query)
    message(paste("Raw data is saved in ", filepath))

    message("--------------------")
    message("oo Preparing data")
    message("--------------------")
    prep_CNV <- GDCprepare(query)
    prep_CNV <- unique(prep_CNV)

    # Generates output
    message("---------------------------")
    message("ooo Generating output...")
    message("---------------------------")


    write.csv(CNV_files, here::here("TCGA_results", project, automataID, "file_information",
                                    sprintf("CNV_files_%s_%s.csv", project, automataID)))
    message(paste("File information has been saved in",
                  here::here("TCGA_results", project, automataID, "file_information",
                             sprintf("CNV_files_%s_%s.csv", project, automataID))))

    write.csv(prep_CNV, here::here("TCGA_results", project, automataID, "TCGA_data",
                                   sprintf("CNV_data_%s_%s.csv", project, automataID)))
    message(paste("CNV data has been saved in",
                  here::here("TCGA_results", project, automataID, "TCGA_data",
                             sprintf("CNV_data_%s_%s.csv", project, automataID))))
  } else{
    message("o Data has been already downloaded and processed.")
    message("o Skipping...")
  }

}

#' Download SNV data from the TCGA
#'
#' Downloads a MAF file for each workflow (MuSE, MuTect2, VarScan2, SomaticSniper) and merges them
#' together keeping the SNPs present in the four of them and the indels present in two of them.
#'
#' @param project Name of a TCGA project (eg. TCGA-PRAD or TCGA-BRCA).
#' @param barcodes Vector containing sample barcodes for the cases to download.
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @return Downloads the four MAF files in the GDC folder and the processed data in TCGA_results.
#' @examples
#' downloadSNV(project = "TCGA-CHOL", barcodes = c("TCGA-W5-AA2X-01A", "TCGA-ZU-A8S4-11A"))
#' @export
#' @importFrom TCGAbiolinks GDCquery_Maf
#' @import dplyr
#' @import magrittr

downloadSNV <- function(project, barcodes, automataID){
  message("---------------------------")
  message("~ DOWNLOADING SNV DATA ~")
  message("---------------------------")
  message("")

  if(file.exists(here::here("TCGA_results", project, automataID, "TCGA_data",
                            sprintf("SNV_data_%s_%s.csv", project, automataID))) == FALSE){

    project <- toupper(project)
    createTCGAfolder(project, automataID)

    short_project <- gsub("^.{0,5}", "", project)


    # Downloads the four MAF data
    message("-------------------------")
    message("o Downloading MuSE data")
    message("-------------------------")
    maf_muse <- GDCquery_Maf(tumor = short_project, pipelines = "muse")
    message("")
    message("---------------------------")
    message("o Downloading MuTect2 data")
    message("---------------------------")
    maf_mutect <- GDCquery_Maf(tumor = short_project, pipelines = "mutect2")
    message("")
    message("-----------------------------")
    message("o Downloading VarScan2 data")
    message("-----------------------------")
    maf_varscan <- GDCquery_Maf(tumor = short_project, pipelines = "varscan2")
    message("")
    message("---------------------------------")
    message("o Downloading SomaticSniper data")
    message("---------------------------------")
    maf_ss <-  GDCquery_Maf(tumor = short_project, pipelines = "somaticsniper")

    message("------------------")
    message("o Preparing data")
    message("------------------")

    # Filter the ones with indels first
    filtered_maf_full <- semi_join(maf_varscan, maf_mutect,
                                   by = c("Hugo_Symbol", "Entrez_Gene_Id", "Chromosome", "Start_Position",
                                          "End_Position", "Strand", "Variant_Classification", "Variant_Type",
                                          "Reference_Allele", "Tumor_Seq_Allele1", "Tumor_Seq_Allele2",
                                          "Tumor_Sample_Barcode"))

    # Separate into a file only with indels and one only with snps
    filtered_maf_indels <- filtered_maf_full %>%
      filter(Variant_Type != "SNP")

    filtered_maf_snp <-  filtered_maf_full %>%
      filter(Variant_Type == "SNP")

    # Keeps filtering only by snps
    filtered_maf_snp <- semi_join(filtered_maf_snp, maf_muse,
                                  by = c("Hugo_Symbol", "Entrez_Gene_Id", "Chromosome", "Start_Position",
                                         "End_Position", "Strand", "Variant_Classification", "Variant_Type",
                                         "Reference_Allele", "Tumor_Seq_Allele1", "Tumor_Seq_Allele2",
                                         "Tumor_Sample_Barcode"))

    filtered_maf_snp <- semi_join(filtered_maf_snp, maf_ss,
                                  by = c("Hugo_Symbol", "Entrez_Gene_Id", "Chromosome", "Start_Position",
                                         "End_Position", "Strand", "Variant_Classification", "Variant_Type",
                                         "Reference_Allele", "Tumor_Seq_Allele1", "Tumor_Seq_Allele2",
                                         "Tumor_Sample_Barcode"))

    # And merge all together again
    filtered_maf_full <- bind_rows(filtered_maf_snp, filtered_maf_indels)

    maf_snvs <- filter(filtered_maf_full, substr(Tumor_Sample_Barcode, 1, 16) %in% barcodes)

    write.csv(maf_snvs, here::here("TCGA_results", project, automataID, "TCGA_data",
                                   sprintf("SNV_data_%s_%s.csv", project, automataID)))
    message(paste("SNV data has been saved in",
                  here::here("TCGA_results", project, automataID, "file_information",
                             sprintf("SNV_data_%s_%s.csv", project, automataID))))

  } else{
    message("o Data has been already downloaded and processed.")
    message("o Skipping...")
  }
}


#' Download methylation data from the TCGA
#'
#' Downloads metylation data for the specified project in the TCGA.
#'
#' @param project Name of the TCGA project data to download (eg. TCGA-PRAD or TCGA-BRCA).
#' @param barcodes Vector containing sample barcodes for the cases to download.
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @return Downloads all the raw data in a folder and creates a  metylation matrix saved in TCGA-results
#' @examples
#' downloadMeth(project = "TCGA-CHOL", barcodes = c("TCGA-W5-AA2X-01A", "TCGA-ZU-A8S4-11A"))
#' @note Barcodes need to contain the sample code (01A or 11A among others). An easy way to get barcodes
#' is using the function getBarcodes.
#' @export


downloadMeth <- function(project, barcodes, automataID){

  message("----------------------------------")
  message("~ DOWNLOADING METHYLATION DATA ~")
  message("----------------------------------")

  if(file.exists(here::here("TCGA_results", project, automataID, "TCGA_data",
                            sprintf("meth_matrix_%s_%s.csv", project, automataID))) == FALSE){

    # Creates folder to save the data
    project <- toupper(project)
    createTCGAfolder(project, automataID)

    # Downloads the data
    if(project == "TCGA-LAML"){

      query <- GDCquery(project = project,
                        data.category = "DNA Methylation",
                        barcode = barcodes,
                        platform = "Illumina Human Methylation 450")
    } else {
      query <- GDCquery(project = project,
                        data.category = "DNA Methylation",
                        barcode = barcodes)
    }




    # Generates a table data of the found files
    meth_files <- getResults(query)

    # Downloads the files (it will create a series of subfolders in the working directory automatically)
    message("")
    message("")
    message("--------------------")
    message("o Downloading data")
    message("--------------------")
    GDCdownload(query,
                method = "api",
                files.per.chunk = 15)

    filepath <- getQueryFilePaths(query)
    message(paste("Raw data is saved in ", filepath))

    message("--------------------")
    message("oo Preparing data")
    message("--------------------")
    message("ooo Extracting methylation matrix")
    meth_matrix <- GDCprepare(query = query, summarizedExperiment = FALSE)

    # Generates output
    message("-------------------------")
    message("ooo Generating output...")
    message("-------------------------")


    write.csv(meth_files, here::here("TCGA_results", project, automataID, "file_information",
                                     sprintf("meth_files_%s_%s.csv", project, automataID)))
    message(paste("File information has been saved in",
                  here::here("TCGA_results", project, automataID, "file_information",
                             sprintf("meth_files_%s_%s.csv", project, automataID))))

    write.csv(meth_matrix, here::here("TCGA_results", project, automataID, "TCGA_data",
                                      sprintf("meth_matrix_%s_%s.csv", project, automataID)))
    message(paste("meth data has been saved in",
                  here::here("TCGA_results", project, automataID, "file_information",
                             sprintf("meth_matrix_%s_%s.csv", project, automataID))))


  } else{
    message("o Data has been already downloaded and processed.")
    message("o Skipping...")
  }

}
