#' Create TCGA_results folder
#'
#' Creates the folder TCGA_results to save the data generated by different functions of Automata.
#'
#' @param project Name of a TCGA project (eg. TCGA-PRAD or TCGA-BRCA).
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @return Generates a series of folders and subfolders for each cancer project and automata ID.
#' @import here
#' @example createTCGAFolder("TCGA-CHOL", automataID)

createTCGAfolder <- function(project, automataID){

  message("o Creating folders for the download...")
  ifelse(!dir.exists(here::here("TCGA_results")), dir.create(here::here("TCGA_results")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project)),
         dir.create(here::here("TCGA_results", project)), "Projec folder already exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID)),
         dir.create(here::here("TCGA_results", project, automataID)), "Folder already exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "file_information")),
         dir.create(here::here("TCGA_results", project, automataID, "file_information")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "TCGA_data")),
         dir.create(here::here("TCGA_results", project, automataID, "TCGA_data")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "processed_expression")),
         dir.create(here::here("TCGA_results", project, automataID, "processed_expression")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "lpd_parameters")),
         dir.create(here::here("TCGA_results", project, automataID, "lpd_parameters")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "report_plots")),
         dir.create(here::here("TCGA_results", project, automataID, "report_plots")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "expression_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "expression_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "expression_analysis", "pathway_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "expression_analysis", "pathway_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "methylation_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "methylation_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "methylation_analysis", "pathway_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "methylation_analysis", "pathway_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "cnv_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "cnv_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "cnv_analysis", "pathway_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "cnv_analysis", "pathway_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "snv_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "snv_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "snv_analysis", "pathway_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "snv_analysis", "pathway_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "clinical_analysis")),
         dir.create(here::here("TCGA_results", project, automataID, "clinical_analysis")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "batch_effect")),
         dir.create(here::here("TCGA_results", project, automataID, "batch_effect")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "intersection")),
         dir.create(here::here("TCGA_results", project, automataID, "intersection")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "mutationalSignatures")),
         dir.create(here::here("TCGA_results", project, automataID, "mutationalSignatures")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "dendrogram")),
         dir.create(here::here("TCGA_results", project, automataID, "dendrogram")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report")),
         dir.create(here::here("TCGA_results", project, automataID, "Report")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Barcodes")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Barcodes")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Genes")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Genes")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "LPD_parameters")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "LPD_parameters")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Signatures")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Signatures")), "Folder alrady exists")

  ifelse(!dir.exists(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Overall")),
         dir.create(here::here("TCGA_results", project, automataID, "Report_RMD", "Childs", "Overall")), "Folder alrady exists")


  return("Folders have been created.")

}


#' Get barcodes from the TCGA
#'
#' For a given TCGA cancer project, fetches the cases with available expression data and returns the
#' barcodes assigned to those cases.
#'
#' @param project Name of a TCGA project (eg. TCGA-PRAD or TCGA-BRCA).
#' @param automataID ID assigned to the analysis so the files from different analysis are not mixed.
#' @param microarray Wether to check for available microarray data (in the TCGA legacy) or RNA-seq
#' counts (in the harmonised TCGA). Defaults to FALSE.
#' @return A CSV file containing the participant and sample barcode, the sample typem the workflow (RNA-seq)
#' or microarray), the project and the assigned automata id.
#' @importFrom TCGAbiolinks GDCquery getResults
#' @importFrom dplyr arrange
#' @import here
#' @example getBarcodes(project = "TCGA-CHOL", automataID = "cholData", microarray = FALSE)
#' @export

getBarcodes <- function(project, automataID, microarray = FALSE){

  message("------------------------")
  message("~ CHECKING BARCODES ~")
  message("------------------------")

  message("o Checking if the provided arguments are correct")

  # Checks if the arguments are right
  if(missing(project))
    stop("Need to specify TCGA project")

  if(missing(automataID))
    stop("Need to assign an automataID")

  # Makes the argument less vulnerable to typos
  project <- toupper(project)


  if(microarray == FALSE){

    # 1. Check that the project is correct ---------------------------------------


    list_of_projects <- c("TCGA-BRCA", 'TCGA-GBM', 'TCGA-OV', 'TCGA-LUAD',
                          'TCGA-UCEC', 'TCGA-KIRC', 'TCGA-HNSC', 'TCGA-LGG',
                          'TCGA-THCA', 'TCGA-LUSC', 'TCGA-PRAD', 'TCGA-SKCM',
                          'TCGA-COAD', 'TCGA-STAD', 'TCGA-BLCA', 'TCGA-LIHC',
                          'TCGA-CESC', 'TCGA-KIRP', 'TCGA-SARC', 'TCGA-LAML',
                          'TCGA-ESCA', 'TCGA-PAAD', 'TCGA-PCPG', 'TCGA-READ',
                          'TCGA-TGCT', 'TCGA-THYM', 'TCGA-KICH', 'TCGA-ACC',
                          'TCGA-MESO', 'TCGA-UVM','TCGA-DLBC', 'TCGA-UCS',
                          'TCGA-CHOL')

    if((project %in% list_of_projects) == FALSE) stop(
      "The project name is not valid, please make sure to use one of the following projects:
      TCGA-BRCA, TCGA-GBM, TCGA-OV, TCGA-LUAD, TCGA-UCEC, TCGA-KIRC, TCGA-HNSC, TCGA-LGG, TCGA-THCA,
      TCGA-LUSC,TCGA-PRAD, TCGA-SKCM, TCGA-COAD, TCGA-STAD, TCGA-BLCA, TCGA-LIHC, TCGA-CESC, TCGA-KIRP,
      TCGA-SARC, TCGA-LAML, TCGA-ESCA, TCGA-PAAD, TCGA-PCPG, TCGA-READ, TCGA-TGCT, TCGA-THYM, TCGA-KICH,
      TCGA-ACC, TCGA-MESO, TCGA-UVM, TCGA-DLBC, TCGA-UCS, TCGA-CHOL.")


    # 1.1 Fetches barcodes from the TCGA ---------------------------------------

    query <- GDCquery(project = project,
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      experimental.strategy = "RNA-Seq",
                      workflow.type = "HTSeq - Counts")

  }

  if(microarray == TRUE){

    # 2. Check that the project is correct ---------------------------------------
    list_of_projects = c('TCGA-BRCA', 'TCGA-OV', 'TCGA-COAD', 'TCGA-KIRC',
                         'TCGA-GBM', 'TCGA-LAML', 'TCGA-LUSC', 'TCGA-READ',
                         'TCGA-UCEC', 'TCGA-LUAD', 'TCGA-LGG', 'TCGA-KIRP')


    if((project %in% list_of_projects) == FALSE) stop(
      "The project name is not valid or available for micorarray data.
      Please, switch to RNA-seq data or try one of these projects:
      'TCGA-BRCA', 'TCGA-OV', 'TCGA-COAD', 'TCGA-KIRC', 'TCGA-GBM',
      'TCGA-LAML', 'TCGA-LUSC', 'TCGA-READ', 'TCGA-UCEC', 'TCGA-LUAD',
      'TCGA-LGG', 'TCGA-KIRP")


    # 2.1 Fetches barcodes from the TCGA ---------------------------------------

    query = GDCquery(project = project,
                     data.category = "Gene expression",
                     data.type = "Gene expression quantification",
                     experimental.strategy = "Gene expression array",
                     legacy = TRUE)

  }

  # 3. Generates data frame with samples and participants barcodes -------------


  aliquot <- getResults(query)$cases
  barcodes <- data.frame(Participant = substr(aliquot, 0, 12),
                         Sample = substr(aliquot, 0, 16),
                         geneExpression.aliquot = aliquot)

  barcodes$Type <- ifelse(grepl("01A", barcodes$Sample), "Primary Solid Tumor", "Solid Tissue Normal")
  barcodes$Workflow <- ifelse(microarray == TRUE, "Microarray", "RNA-seq")
  barcodes$Project <- project
  barcodes$automataID <- automataID

  barcodes <- arrange(barcodes, Participant)

  # 4. Saves the output -------------------------------------------------------
  createTCGAfolder(project, automataID)

  message("o Exporting results...")

  write.csv(barcodes, here::here("TCGA_results", project, automataID, "TCGA_data",
                                 sprintf("Barcodes_%s_%s.csv", project, automataID)))
  message(paste("Barcodes has been saved in",
                here::here("TCGA_results", project, automataID, "TCGA_data",
                           sprintf("Barcodes_%s_%s.csv", project, automataID))))

  barcodes

}

